orttraining/orttraining/python/orttraining_pybind_state.cc

```diff
@@ -160,6 +145,34 @@ TrainingConfigurationResult ConfigureSessionForTraining(
   return python_config_result;
 }
 
+#if defined(USE_NCCL)
+void CopyMPIContextToTrainingParameters(const onnxruntime::training::MPIContext& mpi_context,
+  TrainingParameters& parameters) {
+    std::cout << "mpi_context.world_rank: " << mpi_context.world_rank << std::endl;
+    std::cout << "mpi_context.local_rank: " << mpi_context.local_rank << std::endl;
+    std::cout << "mpi_context.world_size: " << mpi_context.world_size << std::endl;
+    std::cout << "mpi_context.local_size: " << mpi_context.local_size << std::endl;
+
+    parameters.local_rank = mpi_context.local_rank;
+    parameters.local_size = mpi_context.local_size;
+    if (parameters.world_rank != mpi_context.world_rank) {
+      if (parameters.world_rank != 0)
+        std::cout << "WARNING: TrainingParameters world_rank is not correct, tuned automatically to " << mpi_context.world_rank << std::endl;
+      parameters.world_rank = mpi_context.world_rank;
+    }
+    if (parameters.world_size != mpi_context.world_size) {
+      if (parameters.world_size != 1)
+        std::cout << "WARNING: TrainingParameters world_size is not correct, tuned automatically to " << mpi_context.world_size << std::endl;
+      parameters.world_size = mpi_context.world_size;
+    }
+}
+
+onnxruntime::training::MPIContext& GetMpiContext() {
+  static onnxruntime::training::MPIContext mpi_context = training::setup_mpi();
```

Comment:
looks like setup_mpi() is called at most once here and shutdown_mpi() is called from TrainingSession.finalize() perhaps a different number of times. is that valid?

Reply:
or perhaps separate it entirely from TrainingSession - i see there are some global functions which make use of it.

Reply:
updated with MPIContext as a singleton

Reply:
that seems fine to me. i'm not sure how compatible a singleton is with all usages of ORT, e.g., as a DLL. @snnn what do you think?

Reply:
See: https://docs.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-best-practices . Please search "You should never perform the following tasks from within DllMain" and determine if any of these things could happen when shutdown_mpi is called. If you're not sure, I suggest you move it to TrainingSession.finalize()

Reply:
updated so that shutdown_mpi is called in TrainingSession.finalize() for _WIN32.